### DISCUSSION TITLE
IIS hosting of CoreWCF

### DISCUSSION THREAD

**Discussion by actionthomas** [CoreWCF/CoreWCF#321] [Q&A]
Hi everyone!

First, let me shout a big thank you to everyone involved in this project, it will make our transition to .net5+ so much simpler!

Right now, we're hosting some of our netTcp WCF services behind IIS/WAS, with port sharing. Are these kinds of scenarios considered to be supported in CoreWCF, or should I plan to move away from IIS?

Regards,
Thomas

**Comment by Grauenwolf** [comment]
Does it need to be on the same port? 

I have one project with original WCF running under IIS where I was using port 443 for my REST communication and a separate port for WCF. That second port wasn't managed by IIS. 

I did this because I wanted my internal services to be able to talk to each other on a port that wasn't exposed to the Internet. Side-stepping any IIS issues was just a bonus.

**Reply by mconnew** [reply]
NetTcp hosted under IIS isn't supported for 2 reasons. First, we're relying on the ASP.NET Core integration for IIS and they don't have any support for activation of non-HTTP protocols. Second, IIS hosting depends TCP port sharing which depends on the named pipe binding to connect to SMSvcHost, so neither of these scenarios are supported. If you are using NetTcp, I would suggest moving away from IIS.  
You could run CoreWCF in a self hosted mode in an IIS app, but it would be directly opening the TCP port when your code to start the service runs so there's no way to have your app idle and automatically launched when a request arrives.

**Reply by Grauenwolf** [reply]
I tried running a NetTcp port in a self hosted mode, and it just threw errors about running it in IIS without IIS integration enabled. Do you have an example of this working? Maybe there's just a "disable IIS check" command that I'm missing.

**Reply by padhumailin** [reply]
I'm trying to host wshttpbinding in IIS. Any working sample code to achieve?

**Reply by mconnew** [reply]
I've opened an issue for a sample to be created for this. But basically forget that you are using CoreWCF and get an asp.net core app running hosted in IIS following the asp.net core documentation. Then add CoreWCF to that app as it's simply middleware running on top of asp.net core. For HTTP based services there's nothing special about the face it's running on top of IIS.
#377

**Reply by spaasis** [reply]
@mconnew Is the situation still that NetTcp under IIS is not supported?

**Reply by mconnew** [reply]
Correct. I'm in the middle of adding support for named pipes and unix domain sockets. I'm doing the client side first and named pipes is technically code complete, but working through package/build system issues right now. After that I'll add the support to CoreWCF (which isn't needed for NetTcp port sharing, but is the service side of the client work). The next thing on my plate is reliable sessions in CoreWCF. After that I'll be working on TCP port sharing, which will use the WCF Core client named pipe implementation I'm currently finishing up. This won't bring IIS TCP Activation support which I suspect will need changes made to the asp.net core module for IIS. I'll work that out after I have TCP port sharing working for self hosted apps. I'm anticipating TCP port sharing work being completed around March timeframe of next year.

**Reply by Grauenwolf** [reply]
What about running it in a IIS process, but with its own ports unrelated to the ones IIS is using?

**Reply by spaasis** [reply]
@mconnew alright, thanks for the info and your hard work!

**Reply by mconnew** [reply]
> What about running it in a IIS process, but with its own ports unrelated to the ones IIS is using?

@Grauenwolf, are you talking about what I [described above](https://github.com/CoreWCF/CoreWCF/discussions/321#discussioncomment-622792) or something different?

**Reply by Grauenwolf** [reply]
I don't think so. 

What I used to do under .NET Framework was open a TCP based WPF port that IIS doesn't know about at all. It was used for administrative functionality so I didn't want it to be exposed to the normal IIS pipeline. 

So I don't think port sharing would be involved here unless IIS demands that it controls every port that an IIS hosted process opens.

**Reply by mconnew** [reply]
I think the short answer is yes you can do exactly what you intend.  

IIS doesn't demand that. The IIS port sharing feature is mostly about starting your IIS application when a request comes in. When you deploy an app to IIS, it's not actually running. When a request comes in, IIS launches the w3wp process to handle the request. IIS handles HTTP requests itself for this capability. With other non-http protocols such as Net.Tcp, it registers with WAS (Windows Activation Service) the URL (including port) that your app is listening on. WAS then communicates with the process which is registered to handle this protocol (SMSvcHost) to tell it to start listening on that port and path. When a connection arrives to SMSvcHost for that port and path, it tells WAS a matching request has arrived. WAS tells IIS to activate (or start) the app. When WCF is started, it connects to SMSvcHost to tell it the port and address it wants requests for. SMSvcHost then says "great, I have this one already waiting for you".  

If you are okay not having all of that mechanism wired up, you can use CoreWCF (and you can do the same with WCF too) to start a service listening on the port itself. IIS doesn't care, neither does WAS or SMSvcHost, as long as it's a unique port number that you are listening on. You miss out on having multiple services using the same port number, and you also miss out on automatically starting your service when a request comes in if it's not running yet.   

This is exactly the scenario I previously mentioned, but it can be a little tricky to follow without sufficient context.

**Comment by rspulini** [comment]
Greetings, congratulations on your incredible initiative to migrate WCF. I'm very excited to use the coreWCF.

I need your help:

I have a traditional WCF service, hosted on IIS, working as a WebService.

Follow the instructions in the link

https://devblogs.microsoft.com/dotnet/corewcf-v1-released/

I configured IIS to support hosting for Asp.NET


Using an example from microsoft, I managed to get the answer "Hello world!". So my IIS is able to host asp.net with coreWCF.

I compiled the sample coreWCF project. I published to a folder. I copied it to the folder in IIS where the application was configured.

But when trying to open the service in the browser, the error 404 appears

I checked the Windows log (event log) there is no error.

Would you help me? I'm using HTTP bindings.

An observation:
In the traditional WCF service, binding configurations were done in web.config. Now the bindings are done in code? I couldn't understand this very well. In this case I use the svc file to configure the service in IIS

Thank you very much.

**Comment by ArvindRaga** [comment]
Could you please anyone help me also to host corewcf application on IIS server. It's working fine on local but once I hosted on IIS server then I got 404 error. Kindly help me.

**Comment by kiranirabatti** [comment]
Any update on hosting CoreWCF in IIS?

**Reply by oskarjf** [reply]
Would be good with an update.  I'm also working on upgrading from WCF to CoreWCF and .Net 7 and need to be able to run the solution within IIS.

**Reply by kiranirabatti** [reply]
Instead of going with CoreWCF, we ended up removing the WCF from the architecture, And it is very light weight now.

**Comment by KumG** [comment]
Same problem here. Should it work with IIS or not ?
